### PR TITLE
Implement bicubic sampling for lightmaps

### LINKED ISSUE #49935: Vulkan: LightmapGI bicubic sampling not implemented (unlike BakedLightmap in `3.x`)
See also https://github.com/godotengine/godot/issues/49968, https://github.com/godotengine/godot/issues/50568 and https://github.com/godotengine/godot/issues/52199.

### Godot version

4.0.dev (659178e85)

### System information

Fedora 34, GeForce GTX 1080 (NVIDIA 465.31)

### Issue description

In the `master` branch, baked lightmaps can only use bilinear sampling instead of bicubic sampling. This results in lower visual quality, especially when using the **Static** (formerly **All**) bake mode:

image

With bicubic sampling, aliasing in the lightmap becomes much less noticeable. Once it's reimplemented, it should also be enabled by default like in `3.x`.

### Steps to reproduce

Look for a lightmap bicubic sampling project setting (as is present in `3.x`).

### Minimal reproduction project

[test_lightmap_normalmap.zip](https://github.com/godotengine/godot/files/6720846/test_lightmap_normalmap.zip)

### PR DESCRIPTION
**Pull request by BlueCube3310**
Supersedes #77912
Closes #49935

Adds support for bicubic sampling in LightmapGI, which results in much smoother-looking baked shadows:
| Bilinear | Bicubic |
|---------|--------|
|bilinear|bicubic|
|bilinear|bicubic|

To do:
- [x] Add a project setting for using bicubic sampling,
- [x] Don't break compatibility with older LightmapGIData files.

Future work:
- Optimize the shader code for Mobile and Compatibility,
- Allow specifying island distance in the UV2 generator,
- Smooth out the shadows with the default denoiser

### CODE REVIEW

**Comment by fire** [comment]
We're unable to approve for merge if the github actions is failing.

https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html

**Comment by jcostello** [comment]
The bicubic sampling is adding artifacts around the edges of the boxes in the second example?

**Comment by BlueCube3310** [comment]
The filtering can now be toggled from the ProjectSettings and the lightmaps are backwards-compatible with older ones, so upgrading should be seamless.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/29b3d9e9e538f0aa8effc8ad8bf19a2915292a89), it works as expected in all rendering methods. Code looks good to me at a glance.

Toggling the project setting works without having to restart the project as well, which is nice for providing graphics options in projects. (This is an upgrade from 3.x where a restart was required when using GLES2, but not GLES3.)

Disabled | Enabled
-|-
Disabled | Enabled

PS: I noticed toggling bicubic filtering works on existing projects with lightmaps that were baked without this PR. They seem to look fine in this scenario, though I haven't looked extensively to see if they suffer more from lightmap bleeding when bicubic filtering is enabled compared to lightmaps baked with this PR.

Should we print a warning when bicubic filtering is enabled and a LightmapGIData lacks texture size information?

Also, note that this PR prevents using lightmaps baked *after* this PR is merged from working in older builds not including this PR. The scene will still load but lightmaps won't display until you bake lightmaps again. It's not a dealbreaker, just worth keeping in mind.

**Comment by BlueCube3310** [comment]
> PS: I noticed toggling bicubic filtering works on existing projects with lightmaps that were baked without this PR. They seem to look fine in this scenario, though I haven't looked extensively to see if they suffer more from lightmap bleeding when bicubic filtering is enabled compared to lightmaps baked with this PR.
> 
> Should we print a warning when bicubic filtering is enabled and a LightmapGIData lacks texture size information?

If the texture size information isn't present in the file, it's taken from the lightmap textures instead. 
The light bleeding is the same, regardless of which version of the file is used. The best way to mitigate it would probably be to increase the padding on the atlas textures in the lightmapper.

File: servers/rendering_server.cpp
Comment (akien-mga) [inline servers/rendering_server.cpp:2477]:
```suggestion
	ClassDB::bind_method(D_METHOD("lightmaps_set_bicubic_filter", "enable"), &RenderingServer::lightmaps_set_bicubic_filter);
```
To match the docs. Might also be worth changing the C++ methods argument from `p_use` to `p_enable` for consistency, but it's not required.

File: doc/classes/LightmapGIData.xml
Comment (AThousandShips) [inline doc/classes/LightmapGIData.xml:20]:
```suggestion
				Adds an object that is considered baked within this [LightmapGIData]. [param texture_size] must match the size of the [i]entire[/i] lightmap texture which is used for bicubic filtering when rendering the lightmap.
```
No break here in English I'd say

**Comment by AThousandShips** [comment]
You will also need to add compatibility binds

File: core/extension/gdextension_compat_hashes.cpp
Comment (dsnopek) [inline core/extension/gdextension_compat_hashes.cpp:455]: This shouldn't be added here. These mappings are for a very specific situation where we wanted to avoid using compatibility methods, however, it looks like this PR is adding a compatibility method.

File: scene/3d/lightmap_gi.h
Comment (AThousandShips) [inline scene/3d/lightmap_gi.h:88]: These two should be in `protected`

File: servers/rendering_server.h
Comment (AThousandShips) [inline servers/rendering_server.h:1398]: This needs a compatibility method bind as well

**Comment by AThousandShips** [comment]
You also need to add the two validation messages with arguments to the API validation file, see [here](https://docs.godotengine.org/en/latest/contributing/development/handling_compatibility_breakages.html)

File: servers/rendering_server.compat.inc
Comment (AThousandShips) [inline servers/rendering_server.compat.inc:50]:
```suggestion
	instance_geometry_set_lightmap(p_instance, p_lightmap, p_lightmap_uv_scale, p_lightmap_slice, Vector2());
```
Explicit defaults

**Review by AThousandShips — APPROVED** [review]

**Comment by BlueCube3310** [comment]
Is there a reason for passing the lightmap's texture size to the shader manually? GLSL has a built-in function specifically for that task (textureSize), which appears to have been supported for a long time (https://registry.khronos.org/OpenGL-Refpages/gl4/html/textureSize.xhtml).

**Comment by Calinou** [comment]
> Edit: It looks like this was done to support GLES2, which doesn't have this function.

Indeed, this was a remnant of GLES2 support in 3.x. I think 3.x GLES3 used `textureSize()` too.

**Comment by clayjohn** [comment]
> > Indeed, this was a remnant of GLES2 support in 3.x. I think 3.x GLES3 used `textureSize()` too.
> 
> 3.x GLES3 still used the same method as GLES2 for sampling. Since both GLES3 and WEBGL2 support textureSize, would it be safe to use that instead?

`textureSize()` has the same cost as doing a texture fetch which makes it very expensive, especially on mobile. We should avoid using it for anything that will impact run time performance.

**Comment by clayjohn** [comment]
I think this looks mostly good. I am a bit concerned about the bleeding though. 

@BlueCube3310 You mention that the bleeding comes from using a low res lightmap. Does that mean it goes away (or is not noticeable) with a higher res lightmap? I wonder if using some additional padding when using bicubic would be enough to remove the bleeding?

**Comment by BlueCube3310** [comment]
Yes, the bleeding isn't present on high-res ones. For low-res the lightmap UVs could probably be unwrapped with more padding, so it doesn't bleed onto itself (which is more common than onto other instances).

**Comment by jcostello** [comment]
> Yes, the bleeding isn't present on high-res ones. For low-res the lightmap UVs could probably be unwrapped with more padding, so it doesn't bleed onto itself (which is more common than onto other instances).

Is there a padding option exposed? I can't remember

**Comment by BlueCube3310** [comment]
The texture sizes are now calculated at runtime and tied to the lightmaps, rather than during baking and to the instances. I'm currently leaving the old implementation untouched in case this one causes issues, but if it works correctly I'll remove it before merging.

**Comment by clayjohn** [comment]
I think this looks mostly good, but it seems we are missing something else. I recreated the scene from the original comment manually in 3.x and in 4.3 and the 3.x version looks way better. I can confirm that adding bicubic helps the 4.x version, but I'm worried it is just helping hide something that is fundamentally wrong. Take a look at the results

| | Linear | bicubic |
|----- | ---------|--------|
| 3.x |  Screenshot from 2024-08-16 16-30-59 | Screenshot from 2024-08-16 16-31-05 |
| This PR | Screenshot from 2024-08-16 16-30-54 | Screenshot from 2024-08-16 16-31-11 |

The textures produced by both are roughly the same size

File: servers/rendering_server.cpp
Comment (clayjohn) [inline servers/rendering_server.cpp:3564]: This should be under the lightmapping section "rendering/lightmapping/"

**Comment by clayjohn** [comment]
Looking through the code, I have a feeling the lightmap bake for static shadows is just significantly lower quality than it was in 3.x. So that shouldn't hold back this PR

**Comment by BlueCube3310** [comment]
Some comparison screenshots from Unreal Sun Temple:

| linear | bicubic |
|-------|--------|
| off1 | on1 |
| off2 | on2 |

**Comment by BlueCube3310** [comment]
I've removed the old implementation, so now full compatibility is kept between different versions.

**Comment by atirut-w** [comment]
> I think this looks mostly good, but it seems we are missing something else. I recreated the scene from the original comment manually in 3.x and in 4.3 and the 3.x version looks way better. I can confirm that adding bicubic helps the 4.x version, but I'm worried it is just helping hide something that is fundamentally wrong.

Wow, that's really something. New issue, then?

**Comment by Calinou** [comment]
> As far as I can tell, when OIDN is disabled in 3.5, the blockiness is about the same as in 4.3. So it seems that OIDN was just nicely smoothing out our lightmaps before in a nice way.

Using a non-zero Angular Distance helps a lot with reducing shadow blockiness (or Size for positional lights). Doing so affects performance for real-time shadows though, so I suggest adding a second property that would only affect the baked light. This way, you can bake high-quality shadows without affecting real-time shadows.

We could perhaps make this work as `min(angular_distance_baked, angular_distance)` and set it to a non-zero default value to improve bake quality of the box.

**Comment by atirut-w** [comment]
> We could perhaps make this work as `min(angular_distance_baked, angular_distance)` and set it to a non-zero default value (probably 1.0 for Angular Distance and 0.1 for Size) to improve bake quality of the box.

How does that compare to documenting the difference in behavior?

**Comment by clayjohn** [comment]
> > As far as I can tell, when OIDN is disabled in 3.5, the blockiness is about the same as in 4.3. So it seems that OIDN was just nicely smoothing out our lightmaps before in a nice way.
> 
> Using a non-zero **Angular Distance** helps a lot with reducing shadow blockiness in bakes (or **Size** for positional lights). Doing so affects performance for real-time shadows though, so I suggest adding a second property that would only affect baked light. This way, you can bake high-quality shadows without affecting real-time shadows. This is essentially how it works in 3.x, where variable penumbra shadows are only available in BakedLightmap, not real-time rendering.
> 
> We could perhaps make this work as `min(angular_distance_baked, angular_distance)` and set it to a non-zero default value (probably 1.0 for Angular Distance and 0.1 for Size) to improve bake quality of the box.
> 
> The bake time increase from baking variable penumbra shadows is pretty low compared to the improvement in quality, so I'd say it's worth doing in almost every situation. I can open a proposal for this if you want.

I would just cast multiple rays when baking direct lights to get some aliasing, it could work essentially the same as variable penumbra, but in addition to offsetting the sample coordinates of the light, you would offset the position within the texel being sampled

**Review by Calinou — COMMENTED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/1bd740d18d714f815486b04bf4c6154ef6c355d9), it mostly works as expected.

**Testing project:** [test_shadowmask.zip](https://github.com/user-attachments/files/16650062/test_shadowmask.zip)

The main issue I noticed is that bicubic sampling in Compatibility works in the editor, but not in the running project. This happens regardless of the project setting's value, and regardless of whether the lightmaps were initially baked while using Compatibility.

### Forward+

Disabled | Enabled
-|-
Screenshot_20240818_181424 png webp | Screenshot_20240818_181408 png webp

### Mobile

Disabled | Enabled
-|-
Screenshot_20240818_181503 png webp | Screenshot_20240818_181451 png webp

### Compatibility

Disabled | Enabled
-|-
Screenshot_20240818_181526 png webp | Screenshot_20240818_181518 png webp

Also, the increase in light leaking when bicubic filtering is enabled is rather noticeable in the scene's indoor areas. We should look into increasing padding by default (e.g. from 2 to 4 pixels) to alleviate this (even if bicubic filtering is disabled on bake, so that you can enable it later without needing to bake again – and provide a runtime setting for users to toggle).

**Comment by BlueCube3310** [comment]
Should be fixed now, turns out I forgot to initialize the sampling settings in Compatibility

File: servers/rendering/renderer_rd/shaders/forward_mobile/scene_forward_mobile_inc.glsl
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/forward_mobile/scene_forward_mobile_inc.glsl:77]:
```suggestion
	vec2 light_texture_size;
	float exposure_normalization;
	float pad;
```

This should come out of the padding that exists already so we aren't making the struct any bigger

File: servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered_inc.glsl:100]:
```suggestion
	vec2 light_texture_size;
	float exposure_normalization;
	float pad;
```

File: servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.h
Comment (clayjohn) [inline servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.h:214]:
```suggestion
		float texture_size[2];
		float exposure_normalization;
		float pad;
```
For future reference, this entire struct should be refactored to only be 12 floats. 

GLSL handles mat3s really awkwardly, so they take 12 floats. Which means we have 3 floats of secret padding. Since we only have 3 additional floats of data, we can shrink the size of the struct, but that's work for another day

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h
Comment (clayjohn) [inline servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.h:242]:
```suggestion
		float texture_size[2];
		float exposure_normalization;
		float pad;
```

File: drivers/gles3/storage/texture_storage.h
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.h:556]:
```suggestion
	Vector3i texture_get_size(RID p_texture) const;
```
We use "size" everywhere else. So I would keep it as size here

File: drivers/gles3/storage/texture_storage.cpp
Comment (clayjohn) [inline drivers/gles3/storage/texture_storage.cpp:1683]:
```suggestion
Vector3i TextureStorage::texture_get_size(RID p_texture) const {
```

File: drivers/gles3/storage/light_storage.cpp
Comment (clayjohn) [inline drivers/gles3/storage/light_storage.cpp:1049]:
```suggestion
	Vector3i light_texture_size = GLES3::TextureStorage::get_singleton()->texture_get_size(lightmap->light_texture);
```

File: drivers/gles3/storage/light_storage.cpp
Comment (clayjohn) [inline drivers/gles3/storage/light_storage.cpp:1050]:
```suggestion
	lightmap->light_texture_size = Vector2i(light_texture_size.x, light_texture_size.y);
```

**Review by clayjohn — CHANGES_REQUESTED** [review]
Looks amazing! I have left a few suggestions to clean things up a bit. But overall it looks fantastic. Once you make my suggestions it should be ready to merge

**Review by clayjohn — APPROVED** [review]
Amazing work! Looks great

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
